p2p: Lazy init some bloom filters; fuzz version handshake

dergoegge opened the pull request:
This adds a fuzzing harness dedicated to the version handshake. To avoid determinism issues, the harness creates necessary components each iteration (addrman, peerman, etc). A harness like this would have easily caught https://bitcoincore.org/en/2024/07/03/disclose-timestamp-overflow/.

As a performance optimization, this PR includes a change to `PeerManager` to lazily initialize various filters (to avoid large unnecessary memory allocations each iteration).

dergoegge committed:
[fuzz] Harness for version handshake

dergoegge committed:
[net processing] Lazily initialize m_recent_rejects

dergoegge committed:
[net processing] Lazily initialize m_recent_rejects_reconsiderable

dergoegge committed:
[net processing] Lazily initialize m_recent_confirmed_transactions

dergoegge commented:
> Are you sure it is completely non-existent? 

The coverage we do have comes from the `process_messages` harness, but the harness doesn't fuzz the handshake it only makes sure that the handshake completes, so that the other messages can be fuzzed.

I'll generate a coverage report to show the difference

dergoegge commented:
http://bitcoind-fuzz.dergoegge.de:8000/bitcoin-30413/harnesses/p2p_handshake/coverage_report/coverage/workdir/bitcoin/src/net_processing.cpp.html#L3702

brunoerg commented:
Concept ACK

kevkevinpal commented:
> The coverage we do have comes from the process_messages harness, but the harness doesn't fuzz the handshake it only makes sure that the handshake completes, so that the other messages can be fuzzed.

Concept ACK [75fc22f](https://github.com/bitcoin/bitcoin/pull/30413/commits/75fc22f7729c6df779162edb620f986b5c811d8b)
I think it makes sense to fuzz the handshake itself

Maybe I'm just not too familiar with fuzz testing but this seems very similar to `process_messages`.
Would it make sense to modify `process_messages` to also test the handshake instead of creating another fuzz target?

morehouse commented on src/test/fuzz/p2p_handshake.cpp:
```diff
@@ -0,0 +1,107 @@
+                        std::chrono::seconds{24h}.count()));
+
+        CSerializedNetMsg net_msg;
+        net_msg.m_type = PickValue(fuzzed_data_provider, ALL_NET_MESSAGE_TYPES);
```
Would it make sense to limit message type to `VERSION` and `VERACK`?  Are other messages valid during the handshake?

dergoegge replied:
There are more message types that are allowed after the version message and before the verack (some of them are also allowed post verack), essentially all of the ones processed in [this block](https://github.com/bitcoin/bitcoin/blob/20ccb30b7af1c30cece2b0579ba88aa101583f07/src/net_processing.cpp#L3894-L4061) (`wtxidrelay`, `sendaddrv2`, `sendtxrcncl`, `sendheaders`, `sendcmpct`).

I used `ALL_NET_MESSAGE_TYPES`, so that we can avoid touching this test if we add more handshake related message types (since we'll add all new messages to that container).

morehouse commented on src/test/fuzz/p2p_handshake.cpp:
```diff
@@ -0,0 +1,107 @@
+            continue;
+        }
+
+        SetMockTime(GetTime() +
+                    fuzzed_data_provider.ConsumeIntegralInRange<int64_t>(
+                        -std::chrono::seconds{10min}.count(), // Allow mocktime to go backwards slightly
+                        std::chrono::seconds{24h}.count()));
```
Why increment the time instead of setting it directly?

dergoegge replied:
My idea was that this might help a fuzzer to more realistically simulate how time would change between messages. Setting it directly would of course also work but I thought it might result in more mutations that are completely bogus (the time isn't attacker controlled so this seemed less useful). Happy to change it, probably doesn't make a huge difference. What do you think?

dergoegge replied:
> If time isn't attacker controlled, how would it go backwards?

My thinking was that this could simulate NTP adjustments that go backwards.

> And how realistic is the range [-10m, +24h]? Naively, 24h seems like a long time between messages in the handshake.

I've changed the upper bound to `TIMEOUT_INTERVAL` (20min) which is the default timeout value used for connections.

morehouse reviewed (commented):
Concept ACK

dergoegge commented on src/net_processing.cpp:
```diff
@@ -884,11 +884,22 @@ class PeerManagerImpl final : public PeerManager
+     * RecentRejectsReconsiderableFilter(). */
     uint256 hashRecentRejectsChainTip GUARDED_BY(cs_main);
 
+    CRollingBloomFilter& RecentRejectsFilter() EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
Measuring exec/s without lazy init:

![execs_no_lazy](https://github.com/user-attachments/assets/d8db9f90-8708-4e2d-bee3-ce990d895a4c)

With:

![execs_lazy](https://github.com/user-attachments/assets/b508fec3-1f68-4e5d-a0aa-14904453855f)

dergoegge commented:
@maflcko @brunoerg review ping :)

mzumsande commented on src/net_processing.cpp:
```diff
@@ -897,7 +897,18 @@ class PeerManagerImpl final : public PeerManager
      * Memory used: 1.3 MB
      */
-    CRollingBloomFilter m_recent_rejects GUARDED_BY(m_tx_download_mutex){120'000, 0.000'001};
+    std::unique_ptr<CRollingBloomFilter> m_recent_rejects GUARDED_BY(m_tx_download_mutex){nullptr};
```
The recently merged `ActiveTipChange` (#30111) should to be updated too (`m_recent_rejects` and `m_recent_rejects_reconsiderable` are directly dereferenced there).

dergoegge replied:
Ha, good catch! I missed this because `reset()` is both a method on `unique_ptr` and `CRollingBloomFilter`.

This seems to be harmless and work out the way it should, i.e. `unique_ptr::reset()` will delete the current filter and on the next call to `RecentRejectsFilter()` it will be re-allocated (which afaict would have the same effect as resetting the filter with `CRollingBloomFilter::reset()`). But obviously, I should change it to `m_recent_rejects->reset()` to call `reset()` on the filter not the `unique_ptr`.

mzumsande replied:
Right, it shouldn't have compiled otherwise with `.reset()` instead of `->reset()`, the explanation makes sense! Funny (almost) bug!

dergoegge replied:
Renamed

dergoegge commented:
Addressed https://github.com/bitcoin/bitcoin/pull/30413#discussion_r1695993732.

brunoerg commented on src/net_processing.cpp:
```diff
@@ -4604,15 +4637,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             unique_parents.erase(std::unique(unique_parents.begin(), unique_parents.end()), unique_parents.end());
 
-            // Distinguish between parents in m_recent_rejects and m_recent_rejects_reconsiderable.
+            // Distinguish between parents in m_recent_rejects and RecentRejectsReconsiderableFilter().
```
Shouldn't it be `RecentRejectsFilter()` or am I missing anything?

```suggestion
            // Distinguish between parents in RecentRejectsFilter() and RecentRejectsReconsiderableFilter().
```

maflcko reviewed (commented):
Sorry for not reviewing this earlier.

I think the title seems to imply that this is a test-only change, but you are also lazy-initializing some fields in the peer manager. What about changing the title to `p2p: Lazy init some bloom filters; fuzz version handshake`.

Left some nits. Feel free to ignore.

dergoegge commented:
@maflcko Took all your nits.

dergoegge committed:
scripted-diff: Rename lazily initialized bloom filters

-BEGIN VERIFY SCRIPT-
sed -i 's/m_recent_confirmed_transactions/m_lazy_recent_confirmed_transactions/g' $(git grep -l 'm_recent_confirmed_transactions')
sed -i 's/m_recent_rejects_reconsiderable/m_lazy_recent_rejects_reconsiderable/g' $(git grep -l 'm_recent_rejects_reconsiderable')
sed -i 's/m_recent_rejects/m_lazy_recent_rejects/g' $(git grep -l 'm_recent_rejects')
-END VERIFY SCRIPT-

edilmedeiros reviewed (commented):
Concept ACK

I'm curious why the filters should be made lazy for this to work.

dergoegge commented:
> I'm curious why the filters should be made lazy for this to work.

It's not required but it makes the harness more performant, i.e. we can run more testcases each second (see https://github.com/bitcoin/bitcoin/pull/30413#discussion_r1681310085).

marcofleon reviewed (commented):
Tested ACK afd237bb5d85923273a69f7b45dc6aae6aa1680e. I compared the coverage  of `net_processing` from this harness to the `process_message` and `process_messages` harnesses to see the differences. This target hits more specific parts of the version handshake. The stability looks good as well, at about 94%.

brunoerg commented:
ACK afd237bb5d85923273a69f7b45dc6aae6aa1680e

Left it running overnight (slight changes since it), coverage and performance looks great.

glozow reviewed (commented):
utACK afd237bb5d85923273a69f7b45dc6aae6aa1680e lazy blooms look ok

brunoerg commented on src/test/fuzz/p2p_handshake.cpp:
```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2020-present The Bitcoin Core developers
```
```suggestion
// Copyright (c) 2024-present The Bitcoin Core developers
```

mzumsande reviewed (commented):
Code Review ACK afd237bb5d85923273a69f7b45dc6aae6aa1680e

The lazy initialization seem like a good thing in itself because it might also save a little bit of memory during IBD when we don't need the filters yet - `m_lazy_recent_rejects` and `m_lazy_recent_rejects_reconsiderable` should only be initialized after getting out of IBD.

Maybe, as a follow-up, it could make sense to skip early when in IBD in https://github.com/bitcoin/bitcoin/blob/66e82dc90c598c9c42ff980693ef5367a845e1d0/src/net_processing.cpp#L4227-L4233 so that a stray tx-inv received during IBD doesn't result in the filter being initialized in `AlreadyHaveTx` (when we wouldn't call `AddTxAnnouncement()` for it anyway)

dergoegge commented on src/test/fuzz/p2p_handshake.cpp:
```diff
@@ -0,0 +1,107 @@
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 100)
+    {
+        CNode& connection = *PickValue(fuzzed_data_provider, peers);
+        if (connection.fDisconnect || connection.fSuccessfullyConnected) {
```
Yes, this is what effectively limits it to testing the handshake messages. There was some discussion on comparing the new harness to `process_messages` here as well: https://github.com/bitcoin/bitcoin/pull/30413#issuecomment-2217865662.

dergoegge commented:
@edilmedeiros I've seen people have similar issues on a Mac, I don't think these are real issues but rather false positives.

You can configure without the `address` and/or `undefined` sanitizer and that should get rid of the errors, but this obviously is not a great solution if you want to actually use those sanitizers. A second option could be to try to use our ubsan suppressions (https://github.com/bitcoin/bitcoin/blob/master/test/sanitizer_suppressions/ubsan) and add the errors you are seeing to it. See the fuzz test runner for how to tell ubsan to use the suppressions file: https://github.com/bitcoin/bitcoin/blob/b8755164cf9ce7159837d11d19065a5fd6d7d26d/test/fuzz/test_runner.py#L24

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/304.
